lib/goal/container/container.ts

```diff
@@ -0,0 +1,249 @@
+import { k8sContainerScheduler } from "./k8s";
+import { runningInK8s } from "./util";
+
+/**
+ * Create and return a container goal with the provided container
+ * specification.
+ *
+ * @param displayName Goal display name
+ * @param registration Goal containers, volumes, cache details, etc.
+ * @return SDM container goal
+ */
+export function containerGoal(displayName: string, registration: ContainerRegistration): Container {
+    return new Container({ displayName }).with(registration);
+}
+
+/**
+ * Ports to expose from container.
+ */
+export interface ContainerPort {
+    /**
+     * Number of port to expose from the container. This must be
+     * a valid port number, 0 < x < 65536.
+     */
+    containerPort: number;
+}
+
+/**
+ * Volumes to mount in container.
+ */
+export interface ContainerVolumeMount {
+    /** Path to mount point of volume. */
+    mountPath: string;
+    /** Name of volume from [[GoalContainer.volumes]]. */
+    name: string;
+}
+
+/**
+ * Simplified container abstraction for goals.
+ */
+export interface GoalContainer {
+    /** Unique name for this container within this goal. */
+    name: string;
+    /** Full Docker image name, i.e., `registry/name:tag`. */
+    image: string;
+    /**
+     * Docker command and arguments.  We call this `args` rather than
+     * `command` because we think k8s got it right.
+     */
+    args?: string[];
+    /**
+     * Docker image entrypoint.  We call this `command` rather than
+     * `entrypoint` because we think k8s got it right.
+     */
+    command?: string[];
+    /**
+     * Environment variables to set in Docker container.
+     */
+    env?: Array<{ name: string, value: string }>;
+    /**
+     * Ports to expose from container.
+     */
+    ports?: ContainerPort[];
+    /**
+     * Volumes to mount in container.
+     */
+    volumeMounts?: ContainerVolumeMount[];
+}
+
+/**
+ * Volumes that containers in goal can mount.
+ */
+export interface GoalContainerVolume {
+    /** Volume to be created from local host file system location. */
+    hostPath: {
+        /** Absolute path on host to volume. */
+        path: string;
+    };
+    /** Unique name of volume, referenced by [[ContainerVolumeMount.name]]. */
+    name: string;
+}
+
+/**
+ * File system location of goal project in containers.
+ */
+export const ContainerProjectHome = "/atm/home";
+
+/**
+ * Specification of containers and volumes for a container goal.
+ */
+export interface GoalContainerSpec {
+    /**
+     * Containers to run for this goal.  The goal result is based on
+     * the exit status of the first element of the `containers` array.
+     * The other containers are considered "sidecar" containers
+     * provided functionality that the main container needs to
+     * function.  The working directory of the first container is set
+     * to [[ContainerProjectHome]], which contains the project upon
+     * which the goal should operate.
+     */
+    containers: GoalContainer[];
```

Comment:
In my POC code, I think this used to be `GoalContainer[]` or a function that - given the `SdmGoalEvent` - returned the GoalContainer[]. Any particular reason on why you separated this into the `callback`?

Would be nice to also type this property as `containers: GoalContainer | GoalContainer[];` for convenience.

Reply:
I prefer the callback for a few reasons: 1. it makes the interface less complicated, 2. it is more extensible, i.e., if we add more properties they can be modified by the callback too without introducing more callback types, 4. it allows us to have more information available in the callback, 5. it is consistent with the k8s stuff.

Personally I am not a fan of the T | T[] and the complication that brings to the interface and the code, but that is just my opinion.

Reply:
> I prefer the callback for a few reasons: 1. it makes the interface less complicated, 2. it is more extensible, i.e., if we add more properties they can be modified by the callback too without introducing more callback types, 4. it allows us to have more information available in the callback, 5. it is consistent with the k8s stuff.

Ok, that seems reasonable. Especially 3. convinces me ;-) 

> Personally I am not a fan of the `T | T[]` and the complication that brings to the interface and the code, but that is just my opinion.

Hmm, what could I do to convince you? 

Briefly looking at the code I think you'd have to change:
`const spec: GoalContainerSpec = _.merge({}, { containers: registration.containers, volumes: registration.volumes },`
to
`const spec: GoalContainerSpec = _.merge({}, { containers: toArray(registration.containers), volumes: registration.volumes },` to support this. Please?